gobblin-cluster/src/main/java/org/apache/gobblin/cluster/GobblinTemporalJobLauncher.java

```diff
@@ -0,0 +1,499 @@
+      waitForJobCompletion();
+      jobRunTimer.stop();
+      LOGGER.info(String.format("Job %s completed", this.jobContext.getJobId()));
+    } finally {
+      // The last iteration of output TaskState collecting will run when the collector service gets stopped
+      this.taskStateCollectorService.stopAsync().awaitTerminated();
+      cleanupWorkingDirectory();
+    }
+  }
+
+  @Override
+  protected void executeCancellation() {
+    LOGGER.info("Cancel temporal workflow");
+  }
+
+  protected void removeTasksFromCurrentJob(List<String> workUnitIdsToRemove) throws IOException, ExecutionException,
+                                                                                    RetryException {
+    LOGGER.info("Temporal removeTasksFromCurrentJob");
+  }
+
+  protected void addTasksToCurrentJob(List<WorkUnit> workUnitsToAdd) throws IOException, ExecutionException,
+                                                                            RetryException {
+    LOGGER.info("Temporal addTasksToCurrentJob");
+  }
+
+  /**
+   * Submit a job to run.
+   */
+  private void submitJobToTemporal(List<WorkUnit> workUnits) throws Exception{
+    Map<String, TaskConfig> taskConfigMap = Maps.newHashMap();
+
+    try (ParallelRunner stateSerDeRunner = new ParallelRunner(this.stateSerDeRunnerThreads, this.fs)) {
+      int multiTaskIdSequence = 0;
+      for (WorkUnit workUnit : workUnits) {
+        if (workUnit instanceof MultiWorkUnit) {
+          workUnit.setId(JobLauncherUtils.newMultiTaskId(this.jobContext.getJobId(), multiTaskIdSequence++));
+        }
+        addWorkUnit(workUnit, stateSerDeRunner, taskConfigMap);
+      }
+
+      Path jobStateFilePath;
+
+      // write the job.state using the state store if present, otherwise serialize directly to the file
+      if (this.stateStores.haveJobStateStore()) {
+        jobStateFilePath = GobblinClusterUtils.getJobStateFilePath(true, this.appWorkDir, this.jobContext.getJobId());
+        this.stateStores.getJobStateStore()
+            .put(jobStateFilePath.getParent().getName(), jobStateFilePath.getName(), this.jobContext.getJobState());
+      } else {
+        jobStateFilePath = GobblinClusterUtils.getJobStateFilePath(false, this.appWorkDir, this.jobContext.getJobId());
+        SerializationUtils.serializeState(this.fs, jobStateFilePath, this.jobContext.getJobState());
+      }
+
+      // Block on persistence of all workunits to be finished.
+      // It is necessary when underlying storage being slow and Helix activate task-execution before the workunit being persisted.
+      stateSerDeRunner.waitForTasks(Long.MAX_VALUE);
+
+      LOGGER.debug("GobblinTemporalJobLauncher.createTemporalJob: jobStateFilePath {}, jobState {} jobProperties {}",
+          jobStateFilePath, this.jobContext.getJobState().toString(), this.jobContext.getJobState().getProperties());
+
+      Path workUnitFilePath = new Path(this.inputWorkUnitDir, this.jobContext.getJobId());
```

Comment:
Upon re-reading this code, I am a bit confused by this actually. There are multiple workunits that should be produced yet we are only fetching 1 of them. I assume the workunit file path would be derived from the workunit state. 

Can you make sure to check on this?